### PR TITLE
Clean up logging in ctrl library

### PR DESCRIPTION
**Pull request by ychen-atlassian**
Our logging here is a bit messed up and super confusing.

Actually I dislike the way we handle zap logging because it's super confusing which logger has which fields. (smallish fix: can we always pass logger in the arguments and not as a member variable of a struct?).

Anyway this PR changes some fields and fixes some empty strings.

The main thing this was dealing with was https://github.com/kubernetes/client-go/issues/60#issuecomment-281747911 - kubernetes wipes the GVK of objects pulled using the rest client.

### CODE REVIEW

File: controlled_resource_handler.go [hunk 1]
```diff
@@ -96,11 +97,10 @@ func (g *ControlledResourceHandler) OnDelete(obj interface{}) {
 // This method may be called with an empty controllerName.
 func (g *ControlledResourceHandler) rebuildControllerByName(logger *zap.Logger, namespace, controllerName, addUpdateDelete string) {
 	if controllerName == "" {
+		logger.Info("Object has no controller, so nothing was enqueued")
 		return
 	}
-	logger.
-		With(logz.ControllerName(controllerName)).
```
Comment (ychen-atlassian) [inline controlled_resource_handler.go]: Turns out, "Controller" is a really bad overused word.

I've found that:

- Controller means "kubernetes controller"
- Controller also means owner
- Controller also is a component in the SharedIndexInformer
- Controller is also the thing in MVC that we do in our APIServers.

:kappa:

File: controlled_resource_handler.go [hunk 2]
```diff
@@ -96,11 +97,13 @@ func (g *ControlledResourceHandler) OnDelete(obj interface{}) {
 // This method may be called with an empty controllerName.
 func (g *ControlledResourceHandler) rebuildControllerByName(logger *zap.Logger, namespace, controllerName, addUpdateDelete string) {
 	if controllerName == "" {
+		logger.Info("Object has no controller, so nothing was enqueued")
 		return
 	}
 	logger.
-		With(logz.ControllerName(controllerName)).
```
Comment (ychen-atlassian) [inline controlled_resource_handler.go]: Controller is a terrible name, it means too many different things, so I renamed this.

Actually in some areas, we were using "ControllerName" as "ObjectName", even though "Controller" in this context refers to "OwnerReference"....

**Review by amckague — APPROVED** [review]

File: controlled_resource_handler.go (hunk 2 above)
Comment (nilebox) [reply controlled_resource_handler.go]: I guess we could just use "owner" as a term for our usage of ownerReference, since we only use if for parent-child relationship, and our objects don't really have multiple "owners".

**Review by amckague — APPROVED** [review]

File: generic_handler.go [hunk 3]
```diff
@@ -4,49 +4,56 @@ import (
...
-	ZapNameField ZapNameField
+	Logger    *zap.Logger
+	WorkQueue WorkQueueProducer
+
+	Gvk schema.GroupVersionKind
 }
 
 func (g *GenericHandler) OnAdd(obj interface{}) {
-	g.add(obj.(meta_v1.Object), "added")
+	logger := g.Logger.With(logz.Operation("added"))
```
Comment (ash2k) [inline generic_handler.go]: Can use the constant here?

btw, why not give all the operations a type (`type Operation string`) and make `logz.Operation()` accept it? Probably move it somewhere... see if it makes sense.

File: generic_handler.go [hunk 4]
```diff
@@ -4,49 +4,56 @@ import (
...
+	g.add(logger, obj.(meta_v1.Object))
 }
 
 func (g *GenericHandler) OnUpdate(oldObj, newObj interface{}) {
-	g.add(newObj.(meta_v1.Object), "updated")
+	logger := g.Logger.With(logz.Operation("updated"))
```
Comment (ash2k) [inline generic_handler.go]: constant?

File: generic_handler.go [hunk 5]
```diff
@@ -4,49 +4,56 @@ import (
...
+	g.add(logger, newObj.(meta_v1.Object))
 }
 
 func (g *GenericHandler) OnDelete(obj interface{}) {
 	metaObj, ok := obj.(meta_v1.Object)
+	logger := g.Logger.With(logz.Operation("deleted"))
```
Comment (ash2k) [inline generic_handler.go]: same

**Review by ash2k — COMMENTED** [review]
The idea was to use distinct fields for controller object name, controller GK, object name object GK. It is not working that way (totally possible)?

I like the change, see the comments.

**Comment by ash2k** [comment]
Sorry for "controller" to be so overloaded, but... that is how it is. We could do what Nail is suggesting.

**Comment by ychen-atlassian** [comment]
@ash2k Yup, I understood that "controller_name" and "controller_gk" was used for the "controlling object". I think because this word was overloaded, sometimes code outside Ctrl was using these fields as "object_name".

I renamed them to "delegate_name" and "delegate_gk" (because the LookupHandler also uses these fields), and then I figured, the only time we would log these fields is *right before* the owner object is enqueued in the Handler, which doesn't appear to be a useful field. (I justified useful fields as something that we could use in Splunk to group logs). I put them into the Sugar'd logger instead and removed those fields.

File: generic_handler.go (hunk 3 above)
Comment (ychen-atlassian) [reply generic_handler.go]: I'll need to restructure this PR a bit to do that - there's going to be an import cycle unless I put `Operation` into the logz package.

File: handlers/generic_handler.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	g.WorkQueue.Add(ctrl.QueueKey{
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	})
+}
+
+func (g *GenericHandler) loggerForObj(logger *zap.Logger, obj meta_v1.Object) *zap.Logger {
+	return logger.With(logz.Namespace(obj),
+		logz.Object(obj),
+		logz.ObjectGk(g.Gvk.GroupKind()))
```
Comment (ash2k) [inline handlers/generic_handler.go:59]: This contradicts the comment on the `GenericHandler` struct - object_gk should be set already, no need to set it here.

File: handlers/lookup_handler.go [hunk 7]
```diff
@@ -13,56 +15,63 @@ import (
...
 }
 
 // loggerForObj returns a logger with fields for a controlled object.
-func (e *LookupHandler) loggerForObj(obj meta_v1.Object) *zap.Logger {
-	return e.Logger.With(logz.Namespace(obj), logz.Object(obj),
-		logz.ObjectGk(obj.(runtime.Object).GetObjectKind().GroupVersionKind().GroupKind()))
+func (e *LookupHandler) loggerForObj(logger *zap.Logger, obj meta_v1.Object) *zap.Logger {
+	return logger.With(logz.Namespace(obj),
+		logz.Object(obj),
+		logz.ObjectGk(e.Gvk.GroupKind()))
```
Comment (ash2k) [inline handlers/lookup_handler.go:79]: Same - object_gk should be set already.

File: process/generic_worker.go [hunk 8]
```diff
@@ -38,7 +39,10 @@ func (g *Generic) processNextWorkItem() bool {
 
 	holder := g.Controllers[key.gvk]
 	logger := g.logger.With(logz.NamespaceName(key.Namespace),
-		logz.ControllerName(key.Name), logz.Iteration(atomic.AddUint32(&g.iter, 1)))
+		logz.ControllerGk(key.gvk.GroupKind()),
```
Comment (ash2k) [inline process/generic_worker.go:42]: We don't need to set both gks every time - IIRC this is done once when the logger is constructed for the `Generic`.

**Review by ash2k — CHANGES_REQUESTED** [review]
I think I missed the GK-setting code initially. I don't think it is needed. Am I misunderstanding this PR?

File: handlers/lookup_handler.go (hunk 7 above)
Comment (ychen-atlassian) [reply handlers/lookup_handler.go:79]: We have a field called "controller_gk" which is always present on all operations associated with that controller, starting from construction of the constroller. These include logs that don't involve processing an object.

"object_gk" is not set by `Generic`. It sets `controller_gk` to identify which controller a log message is associated with, even during the periods where it is not processing an object. For example, this distintion is useful in the LookupHandler because, for example:

1. `controller_gk` could be `ServiceDescriptor`
2. `object_gk` could be `RoleBinding`.

I agree that this distinction is less useful in the `Process` method of the generic_worker because `controller_gk` is always equal to `object_gk`. However, we were using `logz.ControllerName`, which was... confusingly named. I believe this field was used to log ownerreferences. To make this consistent with how the rest of the code behaves, I thought it would be reasonable to include both `object_name` and `object_gk` for *any* processing that happens within ctrl, since that is what we are using `object_name` for.

I renamed `controller_name` to `delegate_name` to be better. I am not sure if you intended for `controller_name` and `object_name` to be different things?

There was another change I did in this PR which was to fix the original `logz.ObjectGk` call, because Kubernetes strips the APIVersion and Kind from objects coming from Watch/List. This means the ResourceEventHandlers need to be passed in the Gvk that they should log.

In summary the changes I made were:

1. Get rid of `controller_name`, because the term is overloaded. This was truly being used in a random unpredictable way. I've replaced it with a `delegate_name`
2. Fix `object_gk` because it was always empty string
3. Made usage of `object_gk` consistent.
4. Cleaned up logging of operation

How do we use this in structured logging? Let's say I want to figure out *everything* that has been constructed or processed for a particular controller. I can query splunk for `controller_gk`. If I wanted to look at only processing I can query for `object_gk`. If I want to find all relevant logs I can query for `object_gk` and `delegate_gk`, which should give me all object updates, as well as the logs of anything which was queued as a result of a lookup.

File: process/generic_worker.go (hunk 8 above)
Comment (ychen-atlassian) [reply process/generic_worker.go:42]: See big comment

File: handlers/lookup_handler.go (hunk 7 above)
Comment (ash2k) [reply handlers/lookup_handler.go:79]: Ok, thanks for reminding me how it all works - I have forgotten =) You are right, it all makes sense. Please remove/amend those misleading comments on handler/etc structs to reflect what is really expected of the injected logger. Thanks!

**Review by ash2k — APPROVED** [review]
